docs/Options.rst
.. include:: .special.rst

.. _Options:

Options List
============

Note that the options are saved to a JSON file. JSON supports string, numeric, and boolean (true|false). 
String options must be in quotes. See `JSON <https://www.json.org/>`_ for details.

Below is an example JSON file that demonstrates every possible option. Note, some settings are only 
applicable given algorithm selection and execution environment e.g., GA and grid options.

.. parsed-literal:: 

    {
        :ref:`"author" <author_options_desc>`: "Charles Robert Darwin",
        :ref:`"project_name" <project_name_options_desc>`: "Delicious armadillos",

        :ref:`"algorithm" <algorithm_options_desc>`: "GA",

        :ref:`"GA" <GA_options_desc>`: {
            :ref:`"elitist_num" <elitist_num_options_desc>`: 2,
            :ref:`"crossover_rate" <crossover_rate_options_desc>`: 0.95,
            :ref:`"mutation_rate" <mutation_rate_options_desc>`: 0.95,
            :ref:`"sharing_alpha" <sharing_alpha_options_desc>`: 0.1,
            :ref:`"selection" <selection_options_desc>`: "tournament",
            :ref:`"selection_size" <selection_size_options_desc>`: 2,
            :ref:`"crossover_operator" <crossover_operator_options_desc>`: "cxOnePoint",
            :ref:`"mutate" <mutate_options_desc>`: "flipBit",
            :ref:`"attribute_mutation_probability" <attribute_mutation_probability_options_desc>`: 0.1,
            :ref:`"niche_penalty" <niche_penalty_options_desc>`: 20
        },

        :ref:`"MOGA" <MOGA_options_desc>`: {
            :ref:`"objectives" <objectives_options_desc>`: 3,
            :ref:`"names" <objective_names_options_desc>`: ["objective 1", "objective 2", "objective 3"],
            :ref:`"constraints" <constraints_options_desc>`: 0,
            :ref:`"partitions" <partitions_options_desc>`: 6,
            :ref:`"crossover" <crossover_options_desc>`: "single",
            :ref:`"crossover_rate" <crossover_rate_options_desc>`: 0.95,
            :ref:`"mutation_rate" <mutation_rate_options_desc>`: 0.95,
            :ref:`"attribute_mutation_probability" <attribute_mutation_probability_options_desc>`: 0.1
        },

        :ref:`"PSO" <PSO_options_desc>`: {
            :ref:`"inertia" <inertia_options_desc>`: 0.4,
            :ref:`"cognitive" <cognitive_options_desc>`: 0.5,
            :ref:`"social" <social_options_desc>`: 0.5,
            :ref:`"neighbor_num" <neighbor_num_options_desc>`: 20,
            :ref:`"p_norm" <p_norm_options_desc>`: 2,
            :ref:`"break_on_no_change" <break_on_no_change_options_desc>`: 5
        },

        :ref:`"search_omega_blocks" <search_omega_blocks_options_desc>`: false,
        :ref:`"search_omega_bands" <search_omega_bands_options_desc>`: false,
        :ref:`"max_omega_band_width" <max_omega_band_width_options_desc>`: 0,
        :ref:`"search_omega_sub_matrix" <search_omega_sub_matrix_options_desc>`: false,
        :ref:`"max_omega_sub_matrix" <max_omega_sub_matrix_options_desc>`: 4,
        :ref:`"individual_omega_search" <individual_omega_search_options_desc>`: true,
        :ref:`"max_omega_search_len" <max_omega_search_len_options_desc>`: 8,

        :ref:`"random_seed" <random_seed_options_desc>`: 11,
        :ref:`"num_parallel" <num_parallel_options_desc>`: 4,
        :ref:`"num_generations" <num_generations_options_desc>`: 6,
        :ref:`"population_size" <population_size_options_desc>`: 4,

        :ref:`"num_opt_chains" <num_opt_chains_options_desc>`: 4,

        :ref:`"exhaustive_batch_size" <exhaustive_batch_size_options_desc>`: 100,

        :ref:`"crash_value" <crash_value_options_desc>`: 99999999,

        :ref:`"penalty" <penalty_options_desc>`: {
            :ref:`"theta" <theta_options_desc>`: 10,
            :ref:`"omega" <omega_options_desc>`: 10,
            :ref:`"sigma" <sigma_options_desc>`: 10,
            :ref:`"convergence" <convergence_options_desc>`: 100,
            :ref:`"covariance" <covariance_options_desc>`: 100,
            :ref:`"correlation" <correlation_options_desc>`: 100,
            :ref:`"condition_number" <condition_number_options_desc>`: 100,
            :ref:`"non_influential_tokens" <non_influential_tokens_options_desc>`: 0.00001
        },

        :ref:`"effect_limit" <effect_limit_options_desc>`: 6,

        :ref:`"downhill_period" <downhill_period_options_desc>`: 2,
        :ref:`"num_niches" <num_niches_options_desc>`: 2,
        :ref:`"niche_radius" <niche_radius_options_desc>`: 2,
        :ref:`"local_2_bit_search" <local_2_bit_search_options_desc>`: true,
        :ref:`"final_downhill_search" <final_downhill_search_options_desc>`: true,

        :ref:`"local_grid_search" <local_grid_search_options_desc>`: true,
        :ref:`"max_local_grid_search_bits" <max_local_grid_search_bits_options_desc>`: 3,

        :ref:`"nmfe_path" <nmfe_path_options_desc>`: "/opt/nm751/util/nmfe75",
        :ref:`"model_run_timeout" <model_run_timeout_options_desc>`: 1200,
        :ref:`"model_run_priority_class" <model_run_priority_class_options_desc>`: "below_normal",

        :ref:`"postprocess" <postprocess_options_desc>`: {
            :ref:`"use_r" <use_r_options_desc>`: true,
            :ref:`"post_run_r_code" <post_run_r_code_options_desc>`: "{project_dir}/simplefunc.r",
            :ref:`"r_timeout" <r_timeout_options_desc>`: 30,
            :ref:`"use_python" <use_python_options_desc>`: true,
            :ref:`"post_run_python_code" <post_run_python_code_options_desc>`: "{project_dir}/../simplefunc_common.py"
        },

        :ref:`"use_saved_models" <use_saved_models_options_desc>`: false,
        :ref:`"saved_models_file" <saved_models_file_options_desc>`: "{working_dir}/models0.json",
        :ref:`"saved_models_readonly" <saved_models_readonly_options_desc>`: false,

        :ref:`"keep_key_models" <keep_key_models_options_desc>`: false,
        :ref:`"keep_best_models" <keep_best_models_options_desc>`: true,
        :ref:`"rerun_key_models" <rerun_key_models_options_desc>`: false,

        :ref:`"rerun_front_models" <rerun_front_models_options_desc>`: false,

        :ref:`"remove_run_dir" <remove_run_dir_options_desc>`: false,
        :ref:`"remove_temp_dir" <remove_temp_dir_options_desc>`: true,

        :ref:`"keep_files" <keep_files_options_desc>`: ["dmp.txt", "posthoc.csv"],
        :ref:`"keep_extensions" <keep_extensions_options_desc>`: ["shk", "coi", "cor", "cov"],

        :ref:`"use_system_options" <use_system_options_options_desc>`: true,

        :ref:`"model_cache" <model_cache_options_desc>`: "darwin.MemoryModelCache",
        :ref:`"model_run_man" <model_run_man_options_desc>`: "darwin.GridRunManager",
        :ref:`"grid_adapter" <grid_adapter_options_desc>`: "darwin.GenericGridAdapter",
        :ref:`"engine_adapter" <engine_adapter_options_desc>`: "nonmem",

        :ref:`"skip_running" <skip_running_options_desc>`: false,

        :ref:`"rscript_path" <rscript_path_options_desc>`: "C:/Program Files/R/R-4.3.1/bin/Rscript.exe",
        :ref:`"nlme_dir" <nlme_dir_options_desc>`: "C:/Program Files/Certara/NLME_Engine",
        :ref:`"gcc_dir" <gcc_dir_options_desc>`: "C:/Program Files/Certara/mingw64",

        :ref:`"working_dir" <working_dir_options_desc>`: "~/darwin/Ex1",
        :ref:`"data_dir" <data_dir_options_desc>`: "{project_dir}/data",
        :ref:`"output_dir" <output_dir_options_desc>`: "{project_dir}/output",
        :ref:`"temp_dir" <temp_dir_options_desc>`: "{working_dir}/temp",
        :ref:`"key_models_dir" <key_models_dir_options_desc>`: "{working_dir}/key_models",
        :ref:`"non_dominated_models_dir" <non_dominated_models_dir_options_desc>`: "{working_dir}/non_dominated_models",

        :ref:`"generic_grid_adapter" <generic_grid_adapter_options_desc>`: {
            :ref:`"python_path" <python_path_options_desc>`: "~/darwin/venv/bin/python",
            :ref:`"submit_search_command" <submit_search_command_options_desc>`: "qsub -b y -cwd -o {project_dir}/out.txt -e {project_dir}/err.txt -N '{project_name}'",
            :ref:`"submit_command" <submit_command_options_desc>`: "qsub -b y -o {results_dir}/{run_name}.out -e {results_dir}/{run_name}.err -N {job_name}",
            :ref:`"submit_job_id_re" <submit_job_id_re_options_desc>`: "Your job (\\\\w+) \\\\(\\".+?\\"\\\\) has been submitted",
            :ref:`"poll_command" <poll_command_options_desc>`: "qstat -s z",
            :ref:`"poll_job_id_re" <poll_job_id_re_options_desc>`: "^\\\\s+(\\\\w+)",
            :ref:`"poll_interval" <poll_interval_options_desc>`: 5,
            :ref:`"delete_command" <delete_command_options_desc>`: "qdel {project_stem}-\*"
        }
    }


Description
-------------

Here is the list of all available options. Note that many of the options have default values and are not required to be specified directly in the options file.

.. _author_options_desc:

* | :opt_name:`author` -- *string*: The author of the project.
  | Aliased as :mono_ref:`{author}<author_alias>`.

.. _project_name_options_desc:

* | :opt_name:`project_name` -- *string*: Name of the project. By default, it is set to the name of the parent folder of the options file.
  | Aliased as :mono_ref:`{project_name}<project_name_alias>`. See also :mono_ref:`{project_stem}<project_stem_alias>`.

.. _algorithm_options_desc:

* :opt_name:`algorithm` :sup:`required` -- *string*: One of :ref:`EX<EX_desc>`, :ref:`GA<GA_desc>`, :ref:`GP<GP_desc>`, :ref:`RF<RF_desc>`, :ref:`GBRT<GBRT_desc>`, :ref:`PSO<PSO_desc>`, :ref:`MOGA<MOGA_desc>`, :ref:`MOGA3<MOGA3_desc>`.

.. _GA_options_desc:

* :opt_name:`GA` -- *JSON*: Options specific to GA. Ignored for all other algorithms.

.. _elitist_num_options_desc:

    * | :opt_name:`elitist_num` -- *positive int*: Number of best models from any generation to carry over, unchanged, to the next generation. Functions like `Hall of Fame <https://deap.readthedocs.io/en/master/api/tools.html#hall-of-fame/>`_  in DEAP. 
      | *Default*: 4

.. _crossover_rate_options_desc:

    * | :opt_name:`crossover_rate` -- *real*: Fraction of mating pairs that will undergo crossover (real 0.0--1.0).
      | *Default*: 0.95

.. _mutation_rate_options_desc:

    * | :opt_name:`mutation_rate` -- *real*: Probability that at least one bit in the genome will be "flipped", 0 to 1, or 1 to 0, (real 0.0--1.0).
      | *Default*: 0.95

.. _sharing_alpha_options_desc:

    * | :opt_name:`sharing_alpha` -- *real*: Parameter of the niche penalty calculation.
      | *Default*: 0.1

.. _selection_options_desc:

    * | :opt_name:`selection` -- *string*: Selection algorithm for GA. Currently only "tournament" is available.
      | *Default*: ``"tournament"``

.. _selection_size_options_desc:

    * | :opt_name:`selection_size` -- *positive int*: Number of "parents" to enter in the selection. 2 is highly recommended, experience with other values is very limited.
      | *Default*: 2

.. _crossover_operator_options_desc:

    * | :opt_name:`crossover_operator` -- *string*: The algorithm for crossover. Only "cxOnePoint" (single point crossover) is available.
      | *Default*: ``"cxOnePoint"``

.. _mutate_options_desc:

    * | :opt_name:`mutate` -- *string*: The algorithm for mutation. Currently only "flipBit" is available.
      | *Default*: ``"flipBit"``

.. _attribute_mutation_probability_options_desc:

    * | :opt_name:`attribute_mutation_probability` -- *real*: Probability of any bit being mutated, (real 0.0--1.0).
      | *Default*: 0.1

.. _niche_penalty_options_desc:

    * | :opt_name:`niche_penalty` -- *positive real*: Used for calculation of the crowding penalty. 
        The niche penalty is calculated by first finding the "distance matrix", the pair-wise 
        `Mikowski distance <https://en.wikipedia.org/wiki/Minkowski_distance>`_ from the present model to all
        other models. The "crowding" quantity is then calculated as the sum of: (distance/niche_radius)**sharing_alpha
        for all other models in the generation for which the Mikowski distance is less than the niche radius.
      | Finally, the penalty is calculated as: exp((crowding--1)*niche_penalty)--1. The objective of using a niche 
        penalty is to maintain diversity of models, to avoid premature convergence of the search by penalizing when models are too 
        similar to other models in the current generation.
      | *Default*: 20

.. _MOGA_options_desc:

* :opt_name:`MOGA` -- *JSON*: Options specific to MOGA and MOGA3. Ignored for all other algorithms.

.. _objectives_options_desc:

    * | :opt_name:`objectives` -- *positive int*: Number of objectives. Objectives must be provided by :ref:`postprocessing <postprocess_options_desc>` (either R or Python).
      | Applicable only for MOGA3. For MOGA this option is ignored, 2 objectives are used.
      | *Default*: 3

.. _constraints_options_desc:

    * | :opt_name:`constraints` -- *positive int*: Number of constraints. See https://pymoo.org/constraints/index.html
      | Constraints must be provided by :ref:`postprocessing <postprocess_options_desc>` (either R or Python).
      | Applicable only for MOGA3, ignored for MOGA.
      | *Default*: 0

.. note::
   For MOGA3 the postprocessing script must return 2 lists: the first list contains objectives, the second -- constraints. If there are no constraints, the second list must be empty.

.. _objective_names_options_desc:

    * | :opt_name:`names` -- *list of strings*: List of names of the objectives. Must be the size of :ref:`objectives<objectives_options_desc>`. If empty or of a different size, generic names will be used.
      | The names are used only in results.csv.
      | Applicable only for MOGA3, ignored for MOGA.

.. _partitions_options_desc:

    * | :opt_name:`partitions` -- *positive int*: Number of partitions. See https://pymoo.org/misc/reference_directions.html
      | Applicable only for MOGA3, ignored for MOGA.
      | *Default*: 12

.. _crossover_options_desc:

    * | :opt_name:`crossover` -- *string*: Crossover algorithm for MOGA. When set to single, SinglePointCrossover is used. Otherwise, TwoPointCrossover. See https://pymoo.org/operators/crossover.html#Point-Crossover
      | *Default*: ``"single"``

.. _PSO_options_desc:

* :opt_name:`PSO` -- *JSON*: Options specific to PSO. Ignored for all other algorithms.

.. _inertia_options_desc:

    * | :opt_name:`inertia` -- *real*: Particle coordination movement as it relates to the previous velocity. Commonly denoted as :math:`\\w`.
      | *Default*: 0.4

.. _cognitive_options_desc:

    * | :opt_name:`cognitive` -- *real*: Particle coordination movement as it relates to it's own best known position. Commonly denoted as :math:`c_1`.
      | *Default*: 0.5

.. _social_options_desc:

    * | :opt_name:`social` -- *real*: Particle coordination movement as it relates to current best known position across all particles. Commonly denoted as :math:`c_2`.
      | *Default*: 0.5

.. _neighbor_num_options_desc:

    * | :opt_name:`neighbor_num` -- *positive int*: Number of neighbors that any particle interacts with to determine the social component of the velocity of the next step.
      | Smaller number of neighbors results in  a more thorough search (as the neighborhoods tend to move more independently, allowing the swarm to cover a larger
      | section of the total search space) but will converge more slowly.
      | *Default*: 20

.. _p_norm_options_desc:

    * | :opt_name:`p_norm` -- *positive int*: Minkowski p-norm to use. A Value of 1 is the sum-of-absolute values (or L1 distance) while 2 is the Euclidean (or L2) distance.
      | *Default*: 2

.. _break_on_no_change_options_desc:

    * | :opt_name:`break_on_no_change` -- *positive int*: Number of iterations used to determine whether the optimization has converged.
      | *Default*: 5

.. _search_omega_blocks_options_desc:

* | :opt_name:`search_omega_blocks` -- *boolean*: Set to ``true`` to search omega blocks. (Similar to ``search_omega_bands``, but for NLME)
  | *Default*: false

.. _search_omega_bands_options_desc:

* | :opt_name:`search_omega_bands` -- *boolean*: Set to ``true`` to search omega bands.
  | *Default*: false

.. _max_omega_band_width_options_desc:

* | :opt_name:`max_omega_band_width` -- *positive int*: Maximum size of omega band to use in search.
  | *Default*: 0

.. _search_omega_sub_matrix_options_desc:

* | :opt_name:`search_omega_sub_matrix` -- *boolean*: Set to ``true`` to search omega sub matrix.
  | *Default*: false

.. _max_omega_sub_matrix_options_desc:

* | :opt_name:`max_omega_sub_matrix` -- *positive int*: Maximum size of sub matrix to use in search.
  | *Default*: 4

.. _individual_omega_search_options_desc:

* | :opt_name:`individual_omega_search` -- *boolean*: If ``true``, every omega search block will be handled individually: each block will have a separate gene and max omega search length (either calculated automatically or set explicitly with :mono_ref:`max_omega_search_len<max_omega_search_len_options_desc>`).
  | When ``individual_omega_search`` is set to ``false``, the omega search will be performed uniformly, that is, all search blocks will have the same pattern of block omegas.
  | Only search blocks placed directly in the template file can be calculated individually. If any search block is found in tokens, ``individual_omega_search`` is reset to ``false``.
  | *Default*: true

.. _max_omega_search_len_options_desc:

* | :opt_name:`max_omega_search_len` -- *int [2, 16]*: Maximum amount of omegas in a single omega search block. If not set, it will be calculated automatically.

.. _random_seed_options_desc:

* | :opt_name:`random_seed` -- *positive int*: A seed value for random number generator. Used by all machine learning algorithms.
  | The random_seed is also used to generate off-diagonal estimates when :ref:`Searching Omega Structure<omega_search_usage_target>`, regardless if using one of the machine learning algorithms or performing an :ref:`Exhaustive Search<EX_desc>`.

.. _num_parallel_options_desc:

* | :opt_name:`num_parallel` -- *positive int*: Number of models to execute in parallel, i.e., how many threads to create to handle model runs.

docs/Releases.rst
.. include:: .special.rst

######################
Releases
######################

**********************
pyDarwin-Certara 3.0.0
**********************

What's New
====================

* Python 3.12 support.
* Additional support for `RDarwin <https://certara.github.io/R-Darwin/articles/release_notes.html>`_ .

**********************
pyDarwin-Certara 2.0.1
**********************

Issues Corrected
====================

* Estimated OMEGAS are now used when calculating OMEGA penalty, fixed OMEGAS are ignored.
* Estimated SIGMAS are now used when calculating SIGMA penalty, fixed SIGMAS are ignored.
* Unnecessary memory allocation when calculating size of total search space.


**********************
pyDarwin-Certara 2.0
**********************

What's New
====================

* Added NLME Engine support, including :ref:`Omega Block Search <omega_block_search_target>`.

* Introduced :mono_ref:`{darwin_cmd} <darwin_cmd_alias>` alias for Linux Grid runs.

* Introduced :mono_ref:`search_info <search_info>` command.

* Introduced estimations of number of models and remaining time. The former is shown in the beginning of the search as well as in the ``search_info`` output (except for Exhaustive search), the latter -- in the beginning of every but first iteration.
  Those may be not too accurate since they assume a certain amount of Downhill Search iterations (if it is requested) and that a model run time is on average the same (which is not always the case due to duplicates that sometimes make up a large part of the iteration).

* Added tokens consistency check before running the search.

* Added NONMEM license expiration error to the visible errors.

* Updated and renamed columns in **results.csv**.

* Introduced new model statuses:

    - Clone -- a sibling with the same genotype

    - Twin -- a sibling with the same phenotype

    - Restored -- a model run picked from the cache, when the cache was loaded form a file and the model was picked for the first time

    - Cache -- a model run picked from the cache in all other cases

* Improved :ref:`Omega Structure Search <omega_search_usage_target>`, e.g.:

    - added :ref:`individual omega structure search <individual_omega_search_options_desc>`

    - increased number of possible omega block patterns

    - reduced number of model runs by adding the omega structure to the model phenotype, detecting twins, and looking for duplicates in model cache by phenotype

* Introduced key models retention: best models from every iteration :ref:`can be saved <keep_key_models_options_desc>` with all the necessary output in a :ref:`separate folder <key_models_dir_options_desc>` for further analysis.

* Changed :ref:`Final Downhill <final_downhill_search_options_desc>` and :ref:`2-bit search <Local Two bit Search>` iteration names.

* Removed unnecessary 0 generation from :ref:`Genetic Algorithm <GA_desc>`.

* Reorganized examples folder, added NONMEM and :mono:`NLME subfolders`.

* If :mono_ref:`random_seed <random_seed_options_desc>` is not set in the options file it will be initialized with a random number.

* :mono_ref:`working_dir <working_dir_options_desc>` must be an absolute path.

Issues Fixed
====================

* Removed commas and new lines from translation and runtime messages so they won't break CSV structure anymore.

* Corrected calculation of 'Number of unique models to best model'.


**********************
pyDarwin-Certara 1.1.1
**********************

What's New
====================

* The condition_number penalty is now added to the fitness value for the case when the covariance step is successful and condition_number > 1000.
  Previously, the condition_number penalty was added to fitness value only for the case when covariance was unsuccessful or not requested.

Issues Fixed
====================

* An issue was corrected where the correlation penalty does not get added to the resulting fitness value for cases when it should.


********************
pyDarwin-Certara 1.1
********************

What's New
====================

* pyDarwin now supports additional options for :ref:`Searching Omega Structure<omega_search_usage_target>`, including:

    * Band matrix

    * Submatrices

* pyDarwin now supports usage of the :ref:`Particle Swarms Optimization (PSO)<PSO_desc>` algorithm to search for a global optimal solution in the candidate model search space.


Issues Fixed
====================

* An issue has been fixed that could cause the search to fail if the covariance step ($COV) was not specified in template.txt.

********************
pyDarwin-Certara 1.0
********************

Initial release of pyDarwin-Certara offers the Python module `darwin` to
perform a search over a candidate model space using one of the following
machine learning :ref:`algorithms<The algorithms>`:

* :ref:`Genetic Algorithm (GA)<GA_desc>`
* :ref:`Gradient Boosted Random Trees (GBRT)<GBRT_desc>`
* :ref:`Random Forest (RF)<RF_desc>`
* :ref:`Gaussian Process (GP)<GP_desc>`

Users can alternatively select the :ref:`Exhaustive Search (EX) <EX_desc>` to execute
all potential candidate models in a given search space without machine learning
optimization.

Two primary functions to execute the search have been made available:

.. code:: python

    python -m darwin.run_search <template_path> <tokens_path> <options_path>

To execute, call the ``darwin.run_search`` function and provide the paths to the following files as arguments:

1. :ref:`Template file <template_file_target>` (e.g., template.txt) - basic shell for NONMEM control files
2. :ref:`Tokens file <tokens_file_target>` (e.g., tokens.json) - json file describing the dimensions of the search space and the options in each dimension
3. :ref:`Options file <options_file_target>` (e.g., options.json) - json file describing algorithm, run options, and post-run penalty code configurations.


Alternatively, you may execute the :ref:`darwin.run_search_in_folder <darwin.run_search_in_folder>` function,

src/darwin/options.py
import os
import sys
import re
import json
import subprocess
import pathlib
import random

import darwin.utils as utils

from .DarwinError import DarwinError

from darwin.Log import log

_default_penalty = {
    'theta': 10,
    'omega': 10,
    'sigma': 10,
    'convergence': 100,
    'covariance': 100,
    'correlation': 100,
    'condition_number': 100,
    'non_influential_tokens': 0.00001
}

_default_GA = {
    'crossover_rate': 0.95,
    'elitist_num': 4,
    'mutation_rate': 0.95,
    'attribute_mutation_probability': 0.1,
    'mutate': 'flipBit',
    'niche_penalty': 20,
    'selection': 'tournament',
    'selection_size': 2,
    'sharing_alpha': 0.1,
    'crossover_operator': 'cxOnePoint'
}

_default_MOGA = {
    'objectives': 3,
    'names': [],
    'constraints': 0,
    'partitions': 12,
    'crossover': 'single',
    'crossover_rate': 0.95,
    'mutation_rate': 0.95,
    'attribute_mutation_probability': 0.1
}

_default_PSO = {
    'elitist_num': 4,
    'cognitive': 0.5,
    'social': 0.5,
    'inertia': 0.4,
    'neighbor_num': 20,  # must check if < pop_size
    'p_norm': 2,
    'break_on_no_change': 5
}


def _get_mandatory_option(opts: dict, name, for_what=None):
    res = opts.get(name)

    if res is None:
        err = f'{name} is mandatory'

        if for_what:
            err += f' for {for_what}'

        raise DarwinError(err)

    return res


def _get_priority_class(opts: dict):
    if sys.platform != "win32":
        return 0

    priorities = {
        'idle': subprocess.IDLE_PRIORITY_CLASS,
        'below_normal': subprocess.BELOW_NORMAL_PRIORITY_CLASS,
        'normal': subprocess.NORMAL_PRIORITY_CLASS,
        'above_normal': subprocess.ABOVE_NORMAL_PRIORITY_CLASS,
        'high': subprocess.HIGH_PRIORITY_CLASS
    }

    priority = str(opts.get('model_run_priority_class', 'below_normal')).lower()
    opts['model_run_priority_class'] = priority

    if priority not in priorities:
        priority = 'below_normal'

    return priorities[priority]


def _load_system_options(opts: dict) -> dict:
    global_opts_file = os.environ.get('PYDARWIN_OPTIONS')

    if global_opts_file:
        if not os.path.exists(global_opts_file):
            log.warn(f'System options file not found: {global_opts_file}')
            return opts

        log.message(f'Loading system options: {global_opts_file}')

        global_opts = json.loads(open(global_opts_file, 'r').read())

        for x, y in global_opts.items():
            if type(y) == dict:
                opts[x] = opts.get(x, {}) | y
            else:
                opts[x] = y

    return opts


class Options:
    def __init__(self):
        self._options = {}

    def __getitem__(self, key):
        return self._options[key]

    def get(self, key, default):
        return self._options.get(key, default)

    def apply_aliases(self, text: str) -> str:
        return utils.apply_aliases(text, self.aliases)

    def _init_options(self, options_file: str, folder):
        opts = json.loads(open(options_file, 'r', encoding='utf-8').read())

        if opts.get('use_system_options', True):
            opts = _load_system_options(opts)

        self._options = opts

        self.algorithm = _get_mandatory_option(opts, 'algorithm')

        self.engine_adapter = opts.get('engine_adapter', 'nonmem')
        self.model_cache_class = opts.get('model_cache', 'darwin.MemoryModelCache')
        self.model_run_man = opts.get('model_run_man', 'darwin.LocalRunManager')
        self.grid_adapter = opts.get('grid_adapter', 'darwin.GenericGridAdapter')

        self.LOCAL_RUN = self.model_run_man == 'darwin.LocalRunManager'

        self.MAX_RAND_SEED = int(opts.get('MAX_RAND_SEED', 31415926))

        try:
            self.random_seed = int(opts.get('random_seed', random.randint(0, self.MAX_RAND_SEED)))
        except ValueError:
            self.random_seed = None

        try:
            self.num_parallel = int(opts.get('num_parallel', 4))
        except ValueError:
            self.num_parallel = 0

        if self.num_parallel < 1:
            raise DarwinError('num_parallel must be a positive integer')

        self.options_file = os.path.abspath(options_file)

        options_file_parent = pathlib.Path(self.options_file).parent

        self.project_name = opts.get('project_name') or options_file_parent.name
        self.project_stem = re.sub(r'\W', '_', self.project_name)

        darwin_home = os.environ.get('PYDARWIN_HOME') or os.path.join(pathlib.Path.home(), 'pydarwin')

        self.project_dir = str(folder or options_file_parent)
        self.working_dir = utils.apply_aliases(opts.get('working_dir'), {'project_dir': self.project_dir})\
            or os.path.join(darwin_home, self.project_stem)

        if not os.path.isabs(self.working_dir):
            raise DarwinError('working_dir must be an absolute path')

        project_dir_alias = {'project_dir': self.project_dir, 'working_dir': self.working_dir}

        self.isMOGA = self.algorithm == 'MOGA' or self.algorithm == 'MOGA3'
        self.isMOGA3 = self.algorithm == 'MOGA3'
        self.isGA = self.algorithm == 'GA'
        self.isPSO = self.algorithm == 'PSO'

        self.data_dir = utils.apply_aliases(opts.get('data_dir'), project_dir_alias) or self.project_dir
        self.output_dir = utils.apply_aliases(opts.get('output_dir'), project_dir_alias) \
            or os.path.join(self.working_dir, 'output')
        self.temp_dir = utils.apply_aliases(opts.get('temp_dir'), project_dir_alias) \
            or os.path.join(self.working_dir, 'temp')
        self.key_models_dir = utils.apply_aliases(opts.get('key_models_dir'), project_dir_alias) \
            or os.path.join(options.working_dir, 'key_models')

        if self.isMOGA:
            self.non_dominated_models_dir = \
                utils.apply_aliases(opts.get('non_dominated_models_dir'), project_dir_alias) \
                or os.path.join(options.working_dir, 'non_dominated_models')

        self.aliases = {
            'project_dir': self.project_dir,
            'working_dir': self.working_dir,
            'data_dir': self.data_dir,
            'output_dir': self.output_dir,
            'temp_dir': self.temp_dir,
            'algorithm': self.algorithm,
            'author': opts.get('author', 'C.R.Darwin'),
            'project_name': self.project_name,
            'project_stem': self.project_stem,
        }

        penalty = opts.get('penalty', {})
        ga = opts.get('GA', {})
        moga = opts.get('MOGA', {})
        pso = opts.get('PSO', {})

        self.penalty = _default_penalty | penalty
        self.GA = _default_GA | ga
        self.MOGA = _default_MOGA | moga
        self.PSO = _default_PSO | pso
        self.use_saved_models = opts.get('use_saved_models', False)
        self.saved_models_file = utils.apply_aliases(opts.get('saved_models_file'), self.aliases)
        self.saved_models_readonly = opts.get('saved_models_readonly', False) and self.use_saved_models

        self.effect_limit = opts.get('effect_limit', -1)
        self.use_effect_limit = self.effect_limit > 0

        if (options.engine_adapter != 'nonmem' or options.algorithm not in ['GA', 'MOGA', 'MOGA3']) and self.use_effect_limit:
            log.warn('Can only use effect_limit with GA/MOGA/MOGA3 and NONMEM, turned off')
            self.use_effect_limit = False

        self.remove_temp_dir = opts.get('remove_temp_dir', False)
        self.remove_run_dir = opts.get('remove_run_dir', False)
        self.no_cleanup = opts.get('no_cleanup', False)

        self.crash_value = opts.get('crash_value', 99999999)

        if self.algorithm in ["GA", "PSO", "GBRT", "RF", "GP", "MOGA", "MOGA3"]:
            self.population_size = _get_mandatory_option(opts, 'population_size', self.algorithm)
            self.num_generations = _get_mandatory_option(opts, 'num_generations', self.algorithm)

        if self.algorithm in ["GBRT", "RF", "GP"]:
            self.num_opt_chains = _get_mandatory_option(opts, 'num_opt_chains', self.algorithm)

        if self.algorithm in ["GA", "PSO", "GBRT", "RF", "GP", "MOGA", "MOGA3"]:
            self.downhill_period = opts.get('downhill_period', -1)
            self.final_downhill_search = opts.get('final_downhill_search', False)
            self.local_2_bit_search = opts.get('local_2_bit_search', False)

            self.local_grid_search = opts.get('local_grid_search', False)
            self.max_local_grid_search_bits = opts.get('max_local_grid_search_bits', 5)

            if self.local_2_bit_search and self.isMOGA:
                log.warn('2-bit search is requested but ignored for MOGA algorithms')

            if self.downhill_period > 0 or self.final_downhill_search:
                self.num_niches = opts.get('num_niches', 2)
                self.niche_radius = float(opts.get('niche_radius', 2.0))
            elif self.algorithm == 'GA':
                self.niche_radius = float(opts.get('niche_radius', 2.0))

        self.model_run_priority = _get_priority_class(opts)
        self.model_run_timeout = int(opts.get('model_run_timeout', 1200))

        if self.engine_adapter == 'nlme':
            self.rscript_path = _get_mandatory_option(opts, 'rscript_path')
        else:
            self.rscript_path = opts.get('rscript_path', None)

        pp_opts = opts.get('postprocess', {})

        self.use_r = pp_opts.get('use_r', False)
        self.use_python = pp_opts.get('use_python', False)

        self.r_timeout = int(pp_opts.get('r_timeout', 90))

        if self.use_r:
            if self.rscript_path is None:
                self.rscript_path = _get_mandatory_option(pp_opts, 'rscript_path')

            rr = utils.apply_aliases(_get_mandatory_option(pp_opts, 'post_run_r_code'), project_dir_alias)

            self.post_run_r_code = os.path.abspath(rr)

        if self.use_python:
            rp = utils.apply_aliases(_get_mandatory_option(pp_opts, 'post_run_python_code'), project_dir_alias)

            self.python_post_process_path = os.path.abspath(rp)

        self.keep_best_models = opts.get('keep_best_models', True)
        self.keep_key_models = opts.get('keep_key_models', False) or self.keep_best_models

        if self.algorithm in ['EX', 'EXHAUSTIVE']:
            if opts.get('keep_best_models', False) or opts.get('keep_key_models', False):
                log.warn('Not keeping key models for Exhaustive search')

            self.keep_best_models = False
            self.keep_key_models = False

        # don't rerun if key models are not kept
        self.rerun_key_models = opts.get('rerun_key_models', False) and self.keep_key_models

        self.rerun_front_models = opts.get('rerun_front_models', True)

        self.TOKEN_NESTING_LIMIT = opts.get('TOKEN_NESTING_LIMIT', 4)

        self.CORRELATION_THRESHOLD = opts.get('CORRELATION_THRESHOLD', 0.95)

        self.fuzzy_eta = opts.get('fuzzy_eta', True)

        self.keep_extensions = opts.get('keep_extensions', [])
        self.keep_files = opts.get('keep_files', [])

        self.skip_running = opts.get('skip_running', False)

    def initialize(self, options_file, folder=None):
        if not os.path.exists(options_file):
            log.error(f"Couldn't find options file '{options_file}', exiting")
            sys.exit()

        log.message(f"Options file found at {options_file}")

        try:
            self._init_options(options_file, folder)
        except DarwinError as error:
            log.error('Option error: ' + str(error))
            sys.exit()
        except Exception as error:
            log.error(str(error))
            log.error(f"Failed to parse JSON options in '{options_file}', exiting")
            sys.exit()


options = Options()
